File: java-version/src/week3/day3/B2805.java
package week3.day3;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

//https://www.acmicpc.net/problem/2805
//나무 자르기
public class B2805 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int max=0;
        int min=0;
        int [] trees= new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            trees[i] = Integer.parseInt(st.nextToken());
            max = Math.max(trees[i], max);
        }

//        Arrays.sort(trees); // 상관없음 붙이든말던

        while (min < max) {  //이분탐색들어감
            int mid = (max+min)/2;
            long sum =0;


            for (int treeH : trees) {
                if (treeH - mid > 0) {
                    sum+= treeH-mid;
                }
            }


            if (sum < m) {
                max = mid;
            } else {
                min = mid+1;
            }


        }
        System.out.println(min - 1);



    }
}
//
//4 11
//        802
//        743
//        457
//        539

File: codingTest/3week-assignment/day3/B2805.py
#https://www.acmicpc.net/problem/2805
#나무자르기



n, m = map(int, input().split())
trees = list(map(int, input().split()))

max_height = max(trees)  # 최대 절단 높이
min_height = 0           # 최소 절단 높이

while min_height < max_height:
    mid_height = (max_height + min_height) // 2
    sum_of_cut = 0

    for tree in trees:
        if tree - mid_height > 0:
            sum_of_cut += tree - mid_height

    if sum_of_cut < m:
        max_height = mid_height
    else:
        min_height = mid_height + 1

print(min_height - 1)
